File: pregnancy_weight.php
# pregnancy_weight.php
#!/usr/bin/env php
<?php

define('DATA_FILE', 'pregnancy_weight.json');

class PregnancyWeight {
    private $height;
    private $preWeight;
    private $entries = [];

    public function __construct() {
        $this->load();
    }

    private function load() {
        if (file_exists(DATA_FILE)) {
            $data = json_decode(file_get_contents(DATA_FILE), true);
            $this->height = $data['height'] ?? null;
            $this->preWeight = $data['pre_weight'] ?? null;
            $this->entries = $data['entries'] ?? [];
        }
    }

    private function save() {
        file_put_contents(DATA_FILE, json_encode([
            'height' => $this->height,
            'pre_weight' => $this->preWeight,
            'entries' => $this->entries
        ], JSON_PRETTY_PRINT));
    }

    public function setHeight($cm) {
        $this->height = $cm;
        $this->save();
        echo "✅ Height set to {$cm} cm\n";
    }

    public function setPreWeight($kg) {
        $this->preWeight = $kg;
        $this->save();
        echo "✅ Pre‑pregnancy weight set to {$kg} kg\n";
    }

    private function bmi($weight) {
        if (!$this->height) return null;
        $h = $this->height / 100.0;
        return $weight / ($h * $h);
    }

    private function bmiCategory($bmiVal) {
        if ($bmiVal < 18.5) return "Underweight";
        if ($bmiVal < 25.0) return "Normal";
        if ($bmiVal < 30.0) return "Overweight";
        return "Obese";
    }

    public function add($weight, $dateStr = null, $note = '') {
        if (!$this->height) {
            echo "❌ Please set your height first: height <cm>\n";
            return;
        }
        $dateStr = $dateStr ?: date('Y-m-d');
        $bmiVal = $this->bmi($weight);
        $entry = [
            'date' => $dateStr,
            'weight' => $weight,
            'note' => $note,
            'bmi' => $bmiVal
        ];
        $this->entries[] = $entry;
        usort($this->entries, function($a, $b) {
            return strcmp($a['date'], $b['date']);
        });
        $this->save();
        echo "✅ Logged {$weight} kg on {$dateStr}\n";
    }

    public function list() {
        if (empty($this->entries)) {
            echo "No entries.\n";
            return;
        }
        echo "\n📋 Weight Entries:\n";
        foreach ($this->entries as $e) {
            $note = $e['note'] ? " – {$e['note']}" : '';
            $bmiStr = $e['bmi'] ? " (BMI: " . round($e['bmi'], 1) . ")" : '';
            echo "  {$e['date']}: " . round($e['weight'], 1) . " kg{$bmiStr}{$note}\n";
        }
    }

    public function stats() {
        if (empty($this->entries)) {
            echo "No entries.\n";
            return;
        }
        $weights = array_column($this->entries, 'weight');
        $total = count($weights);
        $current = $weights[$total - 1];
        $avg = array_sum($weights) / $total;
        $mn = min($weights);
        $mx = max($weights);
        $gain = $this->preWeight ? $current - $this->preWeight : null;

        echo "\n📊 Statistics:\n";
        echo "  Entries: {$total}\n";
        echo "  Current: " . round($current, 1) . " kg\n";
        echo "  Average: " . round($avg, 1) . " kg\n";
        echo "  Min: " . round($mn, 1) . " kg\n";
        echo "  Max: " . round($mx, 1) . " kg\n";
        if ($gain !== null) {
            echo "  Total gain: " . round($gain, 1) . " kg\n";
        }
    }

    public function showBMI() {
        if (empty($this->entries)) {
            echo "No entries.\n";
            return;
        }
        if (!$this->height) {
            echo "❌ Height not set.\n";
            return;
        }
        $current = $this->entries[count($this->entries)-1]['weight'];
        $bmiVal = $this->bmi($current);
        $cat = $this->bmiCategory($bmiVal);
        echo "\n📐 Current BMI: " . round($bmiVal, 1) . " ({$cat})\n";
        echo "BMI Range: 18.5 - 24.9\n";
    }

    public function showGain() {
        if (empty($this->entries)) {
            echo "No entries.\n";
            return;
        }
        if (!$this->preWeight) {
            echo "❌ Pre‑pregnancy weight not set. Use: preweight <kg>\n";
            return;
        }
        $current = $this->entries[count($this->entries)-1]['weight'];
        $first = $this->entries[0]['weight'];
        $gain = $current - $this->preWeight;
        echo "\n📈 Weight Gain:\n";
        echo "  Pre‑pregnancy: " . round($this->preWeight, 1) . " kg\n";
        echo "  First logged: " . round($first, 1) . " kg\n";
        echo "  Current: " . round($current, 1) . " kg\n";
        echo "  Total gain: " . round($gain, 1) . " kg\n";
    }

    public function trend($days = 5) {
        if (count($this->entries) < $days) {
            echo "Need at least {$days} entries for trend.\n";
            return;
        }
        $recent = array_slice($this->entries, -$days);
        $avg = array_sum(array_column($recent, 'weight')) / count($recent);
        echo "\n📈 Trend (last {$days} days):\n";
        echo "  Average weight: " . round($avg, 1) . " kg\n";
        echo "  Recent entries:\n";
        foreach (array_slice($recent, -5) as $e) {
            echo "    {$e['date']}: " . round($e['weight'], 1) . " kg\n";
        }
    }

    public function export($filename = 'weights.csv') {
        if (empty($this->entries)) {
            echo "No entries.\n";
            return;
        }
        $fp = fopen($filename, 'w');
        fputcsv($fp, ['Date', 'Weight (kg)', 'BMI', 'Note']);
        foreach ($this->entries as $e) {
            $bmiStr = $e['bmi'] ? round($e['bmi'], 1) : '';
            fputcsv($fp, [$e['date'], round($e['weight'], 1), $bmiStr, $e['note']]);
        }
        fclose($fp);
        echo "✅ Exported " . count($this->entries) . " entries to {$filename}\n";
    }
}

if ($argc < 2) {
    die("Usage: php pregnancy_weight.php <command> [options]\n");
}
$app = new PregnancyWeight();
$cmd = $argv[1];

switch ($cmd) {
    case 'height':
        if ($argc < 3) die("height <cm>\n");
        $app->setHeight((float)$argv[2]);
        break;
    case 'preweight':
        if ($argc < 3) die("preweight <kg>\n");
        $app->setPreWeight((float)$argv[2]);
        break;
    case 'add':
        if ($argc < 3) die("add <weight> [--date DATE] [--note TEXT]\n");
        $weight = (float)$argv[2];
        $date = null;
        $note = '';
        for ($i=3; $i<$argc; $i++) {
            if ($argv[$i] == '--date' && isset($argv[$i+1])) {
                $date = $argv[++$i];
            }
            if ($argv[$i] == '--note' && isset($argv[$i+1])) {
                $note = $argv[++$i];
            }
        }
        $app->add($weight, $date, $note);
        break;
    case 'list':
        $app->list();
        break;
    case 'stats':
        $app->stats();
        break;
    case 'bmi':
        $app->showBMI();
        break;
    case 'gain':
        $app->showGain();
        break;
    case 'trend':
        $days = 5;
        for ($i=2; $i<$argc; $i++) {
            if ($argv[$i] == '--days' && isset($argv[$i+1])) {
                $days = (int)$argv[++$i];
            }
        }
        $app->trend($days);
        break;
    case 'export':
        $filename = $argv[2] ?? 'weights.csv';
        $app->export($filename);
        break;
    default:
        echo "Unknown command.\n";
}
?>

File: pregnancy_weight.cpp
// pregnancy_weight.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <algorithm>
#include <nlohmann/json.hpp>
#include <getopt.h>

using namespace std;
using json = nlohmann::json;

struct Entry {
    string date;
    double weight;
    string note;
    double bmi;
};

class PregnancyWeight {
private:
    double height = 0;
    double preWeight = 0;
    vector<Entry> entries;
    string dataFile = "pregnancy_weight.json";

    void load() {
        ifstream f(dataFile);
        if (!f.is_open()) return;
        json j;
        f >> j;
        if (j.contains("height")) height = j["height"];
        if (j.contains("preWeight")) preWeight = j["preWeight"];
        if (j.contains("entries")) {
            for (auto& item : j["entries"]) {
                Entry e;
                e.date = item["date"];
                e.weight = item["weight"];
                e.note = item["note"];
                e.bmi = item["bmi"];
                entries.push_back(e);
            }
        }
    }

    void save() {
        json j;
        j["height"] = height;
        j["preWeight"] = preWeight;
        j["entries"] = json::array();
        for (auto& e : entries) {
            j["entries"].push_back({
                {"date", e.date},
                {"weight", e.weight},
                {"note", e.note},
                {"bmi", e.bmi}
            });
        }
        ofstream f(dataFile);
        f << setw(2) << j << endl;
    }

    double bmi(double weight) {
        if (height <= 0) return 0;
        double h = height / 100.0;
        return weight / (h * h);
    }

    string bmiCategory(double bmiVal) {
        if (bmiVal < 18.5) return "Underweight";
        if (bmiVal < 25.0) return "Normal";
        if (bmiVal < 30.0) return "Overweight";
        return "Obese";
    }

    string currentDate() {
        time_t t = time(nullptr);
        char buf[11];
        strftime(buf, sizeof(buf), "%Y-%m-%d", localtime(&t));
        return string(buf);
    }

public:
    PregnancyWeight() { load(); }

    void setHeight(double cm) {
        height = cm;
        save();
        cout << "✅ Height set to " << cm << " cm\n";
    }

    void setPreWeight(double kg) {
        preWeight = kg;
        save();
        cout << "✅ Pre‑pregnancy weight set to " << kg << " kg\n";
    }

    void add(double weight, const string& dateStr, const string& note) {
        if (height <= 0) {
            cout << "❌ Please set your height first: height <cm>\n";
            return;
        }
        string d = dateStr.empty() ? currentDate() : dateStr;
        Entry e;
        e.date = d;
        e.weight = weight;
        e.note = note;
        e.bmi = bmi(weight);
        entries.push_back(e);
        sort(entries.begin(), entries.end(), [](const Entry& a, const Entry& b) {
            return a.date < b.date;
        });
        save();
        cout << "✅ Logged " << weight << " kg on " << d << "\n";
    }

    void list() {
        if (entries.empty()) {
            cout << "No entries.\n";
            return;
        }
        cout << "\n📋 Weight Entries:\n";
        for (auto& e : entries) {
            string note = e.note.empty() ? "" : " – " + e.note;
            string bmiStr = e.bmi > 0 ? " (BMI: " + to_string(e.bmi).substr(0,4) + ")" : "";
            cout << "  " << e.date << ": " << e.weight << " kg" << bmiStr << note << "\n";
        }
    }

    void stats() {
        if (entries.empty()) {
            cout << "No entries.\n";
            return;
        }
        vector<double> weights;
        for (auto& e : entries) weights.push_back(e.weight);
        int total = weights.size();
        double current = weights.back();
        double avg = 0;
        double mn = weights[0], mx = weights[0];
        for (double w : weights) {
            avg += w;
            if (w < mn) mn = w;
            if (w > mx) mx = w;
        }
        avg /= total;
        double gain = preWeight > 0 ? current - preWeight : 0;

        cout << "\n📊 Statistics:\n";
        cout << "  Entries: " << total << "\n";
        cout << "  Current: " << current << " kg\n";
        cout << "  Average: " << avg << " kg\n";
        cout << "  Min: " << mn << " kg\n";
        cout << "  Max: " << mx << " kg\n";
        if (preWeight > 0) cout << "  Total gain: " << gain << " kg\n";
    }

    void showBMI() {
        if (entries.empty()) {
            cout << "No entries.\n";
            return;
        }
        if (height <= 0) {
            cout << "❌ Height not set.\n";
            return;
        }
        double current = entries.back().weight;
        double bmiVal = bmi(current);
        string cat = bmiCategory(bmiVal);
        cout << "\n📐 Current BMI: " << bmiVal << " (" << cat << ")\n";
        cout << "BMI Range: 18.5 - 24.9\n";
    }

    void showGain() {
        if (entries.empty()) {
            cout << "No entries.\n";
            return;
        }
        if (preWeight <= 0) {
            cout << "❌ Pre‑pregnancy weight not set. Use: preweight <kg>\n";
            return;
        }
        double current = entries.back().weight;
        double first = entries.front().weight;
        double gain = current - preWeight;
        cout << "\n📈 Weight Gain:\n";
        cout << "  Pre‑pregnancy: " << preWeight << " kg\n";
        cout << "  First logged: " << first << " kg\n";
        cout << "  Current: " << current << " kg\n";
        cout << "  Total gain: " << gain << " kg\n";
    }

    void trend(int days = 5) {
        if ((int)entries.size() < days) {
            cout << "Need at least " << days << " entries for trend.\n";
            return;
        }
        vector<Entry> recent(entries.end() - days, entries.end());
        double avg = 0;
        for (auto& e : recent) avg += e.weight;
        avg /= recent.size();
        cout << "\n📈 Trend (last " << days << " days):\n";
        cout << "  Average weight: " << avg << " kg\n";
        cout << "  Recent entries:\n";
        int start = max(0, (int)recent.size() - 5);
        for (int i = start; i < (int)recent.size(); i++) {
            cout << "    " << recent[i].date << ": " << recent[i].weight << " kg\n";
        }
    }

    void exportCSV(const string& filename) {
        if (entries.empty()) {
            cout << "No entries.\n";
            return;
        }
        ofstream f(filename);
        f << "Date,Weight (kg),BMI,Note\n";
        for (auto& e : entries) {
            string bmiStr = e.bmi > 0 ? to_string(e.bmi).substr(0,4) : "";
            f << e.date << "," << e.weight << "," << bmiStr << "," << e.note << "\n";
        }
        f.close();
        cout << "✅ Exported " << entries.size() << " entries to " << filename << "\n";
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: pregnancy_weight <command> [options]\n";
        return 1;
    }
    PregnancyWeight app;
    string cmd = argv[1];

    if (cmd == "height") {
        if (argc < 3) { cerr << "height <cm>\n"; return 1; }
        app.setHeight(stod(argv[2]));
    } else if (cmd == "preweight") {
        if (argc < 3) { cerr << "preweight <kg>\n"; return 1; }
        app.setPreWeight(stod(argv[2]));
    } else if (cmd == "add") {
        if (argc < 3) { cerr << "add <weight> [--date DATE] [--note TEXT]\n"; return 1; }
        double weight = stod(argv[2]);
        string dateStr, note;
        for (int i=3; i<argc; i++) {
            if (string(argv[i]) == "--date" && i+1 < argc) dateStr = argv[++i];
            if (string(argv[i]) == "--note" && i+1 < argc) note = argv[++i];
        }
        app.add(weight, dateStr, note);
    } else if (cmd == "list") {
        app.list();
    } else if (cmd == "stats") {
        app.stats();
    } else if (cmd == "bmi") {
        app.showBMI();
    } else if (cmd == "gain") {
        app.showGain();
    } else if (cmd == "trend") {
        int days = 5;
        for (int i=2; i<argc; i++) {
            if (string(argv[i]) == "--days" && i+1 < argc) {
                days = stoi(argv[++i]);
            }
        }
        app.trend(days);
    } else if (cmd == "export") {
        string filename = argc > 2 ? argv[2] : "weights.csv";
        app.exportCSV(filename);
    } else {
        cerr << "Unknown command.\n";
        return 1;
    }
    return 0;
}
